lib/fizzy/span.hpp

```diff
@@ -0,0 +1,55 @@
+// Fizzy: A fast WebAssembly interpreter
+// Copyright 2019-2020 The Fizzy Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <iterator>
+#include <type_traits>
+
+namespace fizzy
+{
+/// The span describes an object that can refer to a contiguous sequence of objects with the first
+/// element of the sequence at position zero.
+///
+/// This is minimal implementation of C++20's std::span:
+/// https://en.cppreference.com/w/cpp/container/span
+/// Only `const T` is supported.
+template <typename T, typename = typename std::enable_if_t<std::is_const_v<T>>>
+class span
```

Comment:
Is it possible to fall back to native const if compiling on C++20 or it is not worth the hassle (it is not any faster/better than this below)?

Reply:
I'm not sure I understand the question.

This is compatible with C++20 and you can just start using `std::span<const uint64_t>`. The C++20 also supports the `std::span<uint64_t>`.

I disabled the usage of the later, because we don't need it. And also because of the implicit conversions involved. E.g. conversion from `const std::vector<uint64_t>&` to `span<uint64_t>` should be disallowed. I prefer not to handle this in the implementation as long as we don't need it.

If by native const you mean `const span<uint64_t>` it is not useful as an argument type (similar to `const int` argument type).
I think about span as a reference type similar to pointers:
- `span<int>` is as `int*`,
- `span<const int>` is as `const int*`,
- `const span<int>` is as `int* const`,
- `const span<const int>` is as `const int* const`.

Reply:
> This is compatible with C++20 and you can just start using std::span<const uint64_t>. The C++20 also supports the std::span<uint64_t>.

The question was can this header be done in a way such that any code using it will either use the native C++20 implementation or this fallback, depending on compiler settings?

Reply:
Yes, this can be done. I can do it here after #405.